src/python/pants/core/util_rules/external_tool.py

```diff
@@ -73,6 +97,15 @@ def my_rule(my_external_tool: MyExternalTool) -> Foo:
     default_version: str
     default_known_versions: List[str]
 
+    # our hard constraints required for this tool, to support the API we expect from it
```

Comment:
Nit: https://www.pantsbuild.org/docs/style-guide#comments
```suggestion
    # Our hard constraints required for this tool, to support the API we expect from it.
```

Reply:
It would be good to split these two fields and document which a user is expected to override (`version_constraints`), and which they should not touch (`constraints`).

Reply:
I got rid of the `constraints` field altogether.. ;)